t/40-settings.t

```diff
@@ -0,0 +1,124 @@
+    MACHINE                  => '64bit',
+    PATCH                    => 1,
+    UPGRADE                  => 1,
+    ISO                      => 'SLE-%VERSION%-%FLAVOR%-%MACHINE%-Build%BUILD%-Media1.iso',
+    WORKER_CLASS             => 'qemu_x86_64',
+    VERSION                  => '15-SP1',
+    FLAVOR                   => 'Installer-DVD',
+    ADDONURL_SDK             => 'ftp://openqa.suse.de/SLE-%VERSION%-SDK-POOL-%ARCH%-Build%BUILD_SDK%-Media1/',
+    DEPENDENCY_RESOLVER_FLAG => 1,
+    DESKTOP                  => 'textmode',
+    DEV_IMAGE                => 1,
+    HDDSIZEGB                => 50,
+    INSTALLONLY              => 1,
+    PATTERNS                 => 'base,minimal',
+    PUBLISH_PFLASH_VARS =>
+      'SLES-%VERSION%-%ARCH%-%BUILD%@%MACHINE%-minimal_with_sdk%BUILD_SDK%_installed-uefi-vars.qcow2',
+    SEPARATE_HOME      => 0,
+    BUILD_HA           => '%BUILD%',
+    BUILD_SES          => '%BUILD%',
+    WORKAROUND_MODULES => 'base,desktop,serverapp,script,sdk',
+};
+
+subtest expand_placeholders => sub {
+    my $str            = OpenQA::Settings::expand_placeholders($settings);
+    my $match_settings = {
+        BUILD_SDK                => '1234',
+        BETA                     => 1,
+        ISO_MAXSIZE              => '4700372992',
+        SHUTDOWN_NEEDS_AUTH      => 1,
+        HDD_1                    => 'SLES-15-SP1-x86_64-1234@64bit-minimal_with_sdk1234_installed.qcow2',
+        PUBLISH_HDD_1            => 'SLES-15-SP1-x86_64-1234@64bit-minimal_with_sdk1234_installed.qcow2',
+        ANOTHER_JOB              => 'SLES-15-SP1-x86_64-1234@64bit-minimal_with_sdk1234_installed.qcow2',
+        ARCH                     => 'x86_64',
+        BACKEND                  => 'qemu',
+        BUILD                    => '1234',
+        BUILD_SLE                => '1234',
+        MACHINE                  => '64bit',
+        PATCH                    => 1,
+        UPGRADE                  => 1,
+        ISO                      => 'SLE-15-SP1-Installer-DVD-64bit-Build1234-Media1.iso',
+        WORKER_CLASS             => 'qemu_x86_64',
+        VERSION                  => '15-SP1',
+        FLAVOR                   => 'Installer-DVD',
+        ADDONURL_SDK             => 'ftp://openqa.suse.de/SLE-15-SP1-SDK-POOL-x86_64-Build1234-Media1/',
+        DEPENDENCY_RESOLVER_FLAG => 1,
+        DESKTOP                  => 'textmode',
+        DEV_IMAGE                => 1,
+        HDDSIZEGB                => 50,
+        INSTALLONLY              => 1,
+        PATTERNS                 => 'base,minimal',
+        PUBLISH_PFLASH_VARS      => 'SLES-15-SP1-x86_64-1234@64bit-minimal_with_sdk1234_installed-uefi-vars.qcow2',
+        SEPARATE_HOME            => 0,
+        BUILD_HA                 => '1234',
+        BUILD_SES                => '1234',
+        WORKAROUND_MODULES       => 'base,desktop,serverapp,script,sdk',
+    };
+    is_deeply($settings, $match_settings, "Settings replaced");
```

Comment:
would you mind making the test example hashes a bit simpler to only have relevant test data included? e.g. one static value, one with variable, one with multiple variable occurences?

Reply:
This hash is copied from a testsuite in OpenQA ( I added a few parameters into it to make it more complex ), so it can trigger the bug in the previous algorithm. The bug happened randomly, so I must make the hash complex.

Reply:
Are you sure the bug was *random*? Can you reference the bug? I don't see how your algorithmical changes would fix a *random* bug unless it is about sorting

Reply:
Yes, I am sure that.  I reproduced this bug with more than 90 percent probability until I made a complex hash.
There is no bug number or reference link. It happened after my previous PR was merged into OpenQA. So the pr was reverted.
For example, {BUILD => '1234', BUILD_HA => '%BUILD%', ISO => 'sle-15-sp1-%BUILD_HA%-x86_64.iso'}, when I loop values %hash, unlucky, I replace ISO before BUILD_HA, as a result the value of ISO is modified to sle-15-sp1-%BUILD%-x86_64.iso. This is the bug in previous PR.

Reply:
Yes, I am sure that. I reproduced this bug with more than 90 percent probability until I made a complex hash.
There is no bug number or reference link. It happened after my previous PR was deployed in OpenQA. So the pr was reverted.
For example, {BUILD => '1234', BUILD_HA => '%BUILD%', ISO => 'sle-15-sp1-%BUILD_HA%-x86_64.iso'}, when I loop values %hash, unlucky, I replace ISO before BUILD_HA, as a result the value of ISO is modified to sle-15-sp1-%BUILD%-x86_64.iso. This is the bug in previous PR.